Add yum::copr resource to handle COPR repositories.

olifre opened the pull request:
COPR (Cool Other Package Repo) is a Fedora project for third-party package repositories, see:
https://copr.fedorainfracloud.org/

#### Pull Request (PR) description
This PR adds support to manage COPR repositories via a new `yum::copr` resource. 
Documentation and test are included. 

#### This Pull Request (PR) fixes the following issues
Fixes #149

traylenator commented on manifests/copr.pp:
```diff
@@ -0,0 +1,79 @@
+#   }
+#
+define yum::copr (
+  String                                 $copr_repo = $title,
```
Variable `$repo` is proably enough since it's already in `yum::copr` type.

olifre replied:
Indeed, I changed that in the last force push :+1: .

traylenator reviewed (commented):
This is a good case for an acceptance test as well I'd say.

traylenator commented on manifests/copr.pp:
```diff
@@ -0,0 +1,79 @@
+  Enum['enabled', 'disabled', 'removed'] $ensure    = 'enabled',
+) {
+  $prereq_plugin = $facts['package_provider'] ? {
+    'dnf'   => 'dnf-plugins-core',
```
I always prefer to make the future, dnf in this case, the default case since it's more obvious what code needs removing once yum is gone from the universe.

olifre replied:
That's a good idea, thanks for the suggestion! I've also addressed this in the last push.

olifre commented:
> This is a good case for an acceptance test as well I'd say.

Indeed it is. I've addressed all comments apart from the acceptance test, will look into this later this week.

bastelfreak commented on manifests/copr.pp:
```diff
@@ -0,0 +1,79 @@
+#   }
+#
+define yum::copr (
+  String                                 $repo   = $title,
```
please add puppet-strings like documentation for the new parameters

olifre replied:
Done in the last force-push, really appreciate all the feedback from review, since I'm not yet used to most of the scaffolding and test infrastructure, and you are pushing me to read into the correct things :+1: .

olifre commented:
I've now fixed the Rubocop offenses in the existing test, and added an acceptance test. I don't have a box at hand (yet) to run the acceptance test, but I think it should work — CI will show ;-).

traylenator commented:
One of the problems with this if you are running with

```puppet
resources{'yumrepo':
   purge => true,
}
```

The repo file will be deleted won't it?

I don't see a good way around that.

olifre commented:
> The repo file will be deleted won't it?

Indeed... 

> I don't see a good way around that.

Same here, at least not as long as we want to use the official way (via `dnf` / `yum`) instead of managing the repo "manually", but then all the magic hidden by the underlying package manager plugins is lost and we'd need to follow all updates on that end. 

Good ideas are very welcome.

olifre commented:
Sorry for the multiple force-pushes, I found a small logic issue in the last commit and fixed it — if there is any good proposal on how to upstream this, fixing / documenting the issue @traylenator pointed out, please let me know.

bastelfreak commented:
@olifre force pushes are fine. Can you take a look at the failing CI jobs?

olifre commented:
> @olifre force pushes are fine. Can you take a look at the failing CI jobs?

Thanls, I think I fixed all CI issues, the remaining issue seems to be a fluke of the GitHub actions runner (`read timeout reached`). I will trigger CI once more with an empty force-push.

olifre commented:
@bastelfreak That did the trick, now all is green :smile: .

olifre commented:
@bastelfreak It seems CI went red again after you re-ran it yesterday, but checking the logs, that seems like an issue with the runners, correct?

traylenator commented:
> One of the problems with this if you are running with
> 
> ```puppet
> resources{'yumrepo':
>    purge => true,
> }
> ```
> 
> The repo file will be deleted won't it?
> 
> I don't see a good way around that.

Given there's no sensible way around this and it only affects particular configurations I would just document it.
If the resources declaration can be adjusted to ignore the `copr` repos based on name or something then makes sense to ask that.

https://forge.puppet.com/modules/crayfishx/purge/readme gives an example in fact.

olifre commented:
> Given there's no sensible way around this and it only affects particular configurations I would just document it.
> If the resources declaration can be adjusted to ignore the `copr` repos based on name or something then makes sense to give that example.

Thanks for confirming!

The names of `yumrepo` resources resulting from COPR commands end up being formed such as:
`copr:copr.fedorainfracloud.org:jdoss:wireguard`
`copr:copr.fedorainfracloud.org:copart:restic`
(tested with RHEL 7 and 8). Since `resources` itself does not seem to allow limiting what is purged, the only way out seems to be to use something like the `purge` module you linked. 

What do you think about the documentation addendum I added in the last commit?

olifre commented:
I have finally come around to also regenerate `REFERENCE.md` and rebase onto current `master`, so things should now be green and mergable.

bastelfreak commented on manifests/copr.pp:
```diff
@@ -0,0 +1,71 @@
+    'dnf'   => 'dnf-plugins-core',
+    default => 'yum-plugin-copr',
+  }
+  stdlib::ensure_packages([$prereq_plugin])
```
please introduce a new param like $manage_prereq_plugin and wrap an if condition around thr function.

olifre replied:
That makes a lot of sense, done and force-pushed.
